### PR TITLE
VS3 to master

### PR DESCRIPTION
**Pull request by Leidse**

### CODE REVIEW

File: src/coordinate3d.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * @file
+ * @brief     This class contains the Coordinate3D Class
+ * @author    Jeroen van Hattem and Jeffrey de Waal
+ * @license   See LICENSE
+ */
+#ifndef COORDINATE3D_HPP
+#define COORDINATE3D_HPP
+
+class Coordinate3D {
+  private:
```
Comment (BIGduzy) [inline src/coordinate3d.hpp]: If the member variables all have getters and setters (without logic like bounds checking) you should just make them public and remove the getters and setters.
Then if a class has only public you should make it a struct.

File: src/robot_arm.cpp [hunk 2]
```diff
@@ -1,19 +1,116 @@
...
+ */
 #include "robot_arm.hpp"
 
-void robot_arm::move(int coordinates[3]) {
-    copy_data(coordinates, goto_coordinates);
+RobotArm::RobotArm() : conn(115200, UARTController::ONE) {
+}
+
+void RobotArm::sendGCodeToArm(const char *command) {
+    if ((hwlib::now_us() / 1000) - startMsSend > 2500) {
```
Comment (BIGduzy) [inline src/robot_arm.cpp]: Add some comments to explain why this is necessary

File: src/robot_arm.cpp [hunk 3]
```diff
@@ -1,19 +1,116 @@
...
+    if (conn.available() > 0 && (hwlib::now_us() / 1000) - startMsReceive > 30) {
+        startMsReceive = hwlib::now_us() / 1000;
 
-    hwlib::cout << "Moving to (" << coordinates[0] << ", " << coordinates[1] << ", " << coordinates[2] << ")" << hwlib::endl;
-    output_pin.set(1);
-    hwlib::wait_ms(1000);
-    output_pin.set(0);
-    hwlib::cout << "Finished moving" << hwlib::endl;
+        for (unsigned int i = 0; i < conn.available(); i++) {
+            hwlib::cout << (char)conn.receive();
```
Comment (BIGduzy) [inline src/robot_arm.cpp]: Use c++ style type casting (static_cast)

File: src/robot_arm.hpp [hunk 4]
```diff
@@ -1,19 +1,132 @@
-#ifndef ROBOT_ARM_HPP
-#define ROBOT_ARM_HPP
+/**
```
Comment (BIGduzy) [inline src/robot_arm.hpp]: Make sure to put your code in the required namespace

File: src/robot_arm.hpp [hunk 5]
```diff
@@ -1,19 +1,132 @@
...
+    /**
+     * @brief Move the arm to the set coordinates
+     *
+     * This function will take an array existing of 3 integer values and an integer that resembles the speed. These parameters
+     * will me transformed into a G-Code. This will be send to the uArm using the uart_connection class.
+     *
+     * @param coordinates : Coordinate3D
+     * @param speed : int
+     */
+    void move(Coordinate3D coordinates, int speed);
```
Comment (BIGduzy) [inline src/robot_arm.hpp]: Make coordinates a const reference

File: src/robot_arm.cpp [hunk 6]
```diff
@@ -1,19 +1,116 @@
...
+    return dest;
+}
+
+char *RobotArm::strcopy(char *dest, const char *src) {
+    char *saved = dest;
+    while (*src) {
+        *dest++ = *src++;
+    }
+    *dest++ = '\0';
+    *dest = 0;
```
Comment (BIGduzy) [inline src/robot_arm.cpp:91]: Why is this necessary?

File: src/robot_arm.hpp [hunk 7]
```diff
@@ -1,19 +1,132 @@
...
+     */
+    void executeAction(Actions action);
+    /**
+     * @brief Send a G-Code to the uArm
+     *
+     * This is the function that sends the G-Code to the uArm.
+     *
+     * @param command
+     */
+    void sendGCodeToArm(const char *command);
```
Comment (BIGduzy) [inline src/robot_arm.hpp:63]: This function could be private, since the user only should use executeAction

File: src/robot_arm.hpp [hunk 8]
```diff
@@ -1,19 +1,132 @@
...
+    void executeAction(const char *action);
+    /**
+     * @brief Determine G-Code for a desired location.
+     *
+     * Determine G-Code for a desired location. This G-Code will be stored in commandBuffer.
+     *
+     * @param coordinates : Coordinate3D
+     * @param speed : int
+     */
+    void determineGCode(Coordinate3D coordinates, int speed);
```
Comment (BIGduzy) [inline src/robot_arm.hpp]: Make coordinates a const reference

**Review by BIGduzy — CHANGES_REQUESTED** [review]
Make sure to read the wiki on how to write the doxygen comments (for params)
The params also do not require a return type since doxygen just takes that from the function.

Overal add more comments inside the functions to explain the logic (E.g. for stradd and strcopy).

**Comment by BIGduzy** [comment]
The conflicts also need to be solved before accepting this pull request.

File: src/robot_arm.cpp (hunk 6 above)
Comment (wvanbreukelen) [reply src/robot_arm.cpp:91]: This is the pure strcpy implementation available in the libc libraries (mentioned in the method description by @jeroenvanhattem).

**Comment by wvanbreukelen** [comment]
As it seems, we need to merge locally to resolve the merge conflict. Maybe @BIGduzy can give his approval (if code is up to standards), so this merge can be performed?

File: src/coordinate3d.hpp [hunk 9]
```diff
@@ -8,12 +8,11 @@
 #define COORDINATE3D_HPP
 
 class Coordinate3D {
```
Comment (BIGduzy) [inline src/coordinate3d.hpp]: Just make this a struct.

**Review by BIGduzy — CHANGES_REQUESTED** [review]
One small change is necessary.

File: src/coordinate3d.hpp (hunk 9 above)
Comment (wvanbreukelen) [reply src/coordinate3d.hpp]: @BIGduzy As we program in C++, and not in C, I think the use of ```class``` would be more appropriate, instead of using ```struct```

**Review by BIGduzy — APPROVED** [review]

**Comment by BIGduzy** [comment]
Approved

**Comment by wvanbreukelen** [comment]
Tests passed, merge conflicts resolved (accepted by @BIGduzy)
